### ISSUE TITLE
GPR_ASSERT(op->context) fails in client_auth_filter

### ISSUE THREAD

**Issue by ohad7** [grpc/grpc#3390]
Hi there
We're experiencing a SIGABRT in auth_start_transport_op (client_auth_filter.c) - I'm assuming that the following assertion is being called and for some reason op->context is null : 

GPR_ASSERT(op->context)

This is the stack trace. the high level call is made to CompletionQueue::AsyncNextInternal

libsystem_kernel.dylib 0x000000019ae0f140 __pthread_kill + 4
libsystem_c.dylib 0x000000019ad82b78 abort + 136
GRPC 0x00000001005efa90 auth_start_transport_op + 584
GRPC 0x00000001005f1058 started_call + 140
GRPC 0x0000000100707038 continue_creating_call + 40
GRPC 0x000000010068904c grpc_maybe_call_delayed_callbacks + 148
GRPC 0x00000001006d04e0 grpc_pollset_work + 68
GRPC 0x00000001005f622c grpc_completion_queue_next + 188

We haven't managed to reproduce the issue - we get it as Apple iOS crash reports from clients in the wild, looking at the logs we see that these happen in a request that's coming right after a timed out request, meaning : 
 - request 1 made
 - request 1 timed out
 - request 2 made
 - crash!

The client is brand new c++ (0.11), On the server we have a Java server with version 0.7.2

**Comment by ohad7** [comment]
Hi again
The issue is reproducible when sending a lot of requests, then shutting down the server and restarting it after some time (1 minute). The client will then fail on the assertion.

Here's an example of the stack when it fails, note that op is all NULLs, and that cancel_with_status is GRPC_STATUS_CANCELED:

![image](https://cloud.githubusercontent.com/assets/9012046/9983167/351b1520-5ffc-11e5-8c1a-d7929bf93c22.png)

@danageva managed to workaround this issue with avoiding entering the assertion if the status is canceled. Here's the code that currently works for us : 

```c++
if (calld->security_context_set == 0 && op->cancel_with_status != GRPC_STATUS_CANCELLED) {
    calld->security_context_set = 1;
    GPR_ASSERT(op->context);
    if (op->context[GRPC_CONTEXT_SECURITY].value == NULL) {
      op->context[GRPC_CONTEXT_SECURITY].value =
      grpc_client_security_context_create();
      op->context[GRPC_CONTEXT_SECURITY].destroy =
      grpc_client_security_context_destroy;
    }
    sec_ctx = op->context[GRPC_CONTEXT_SECURITY].value;
    GRPC_AUTH_CONTEXT_UNREF(sec_ctx->auth_context, "client auth filter");
    sec_ctx->auth_context = GRPC_AUTH_CONTEXT_REF(
                                                  chand->security_connector->base.auth_context, "client_auth_filter");
  }
```

Please let us know whether we're doing the right thing

cc @olympium @dorongoffer @koker @danageva

**Comment by jboeuf** [comment]
Thanks for the bug report. I'm very busy today but I'll try to have a look. Tomorrow latest.

**Comment by jboeuf** [comment]
Hopefully, the fix will make it in time for 0.11.1. Your fix was right but I generalized it a bit more. Thanks!
